src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/InterceptedProjectProperties.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.ProjectSystem.ProjectPropertiesProviders;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties
+{
+    /// <summary>
+    /// An implementation of IProjectProperties that intercepts all the get/set callbacks for each property on
+    /// the given default <see cref="IProjectProperties"/> and passes it to corresponding <see cref="IInterceptingPropertyValueProvider"/>
+    /// to validate and/or transform the property value to get/set.
+    /// </summary>
+    internal sealed class InterceptedProjectProperties : DelegatedProjectPropertiesBase
+    {
+        private readonly ImmutableDictionary<string, IInterceptingPropertyValueProvider> _valueProviders;
+
+        public InterceptedProjectProperties(ImmutableArray<Lazy<IInterceptingPropertyValueProvider>> valueProviders, IProjectProperties defaultProperties)
+            : base(defaultProperties)
+        {
+            Requires.NotNullOrEmpty(valueProviders, nameof(valueProviders));
+
+            var builder = ImmutableDictionary.CreateBuilder<string, IInterceptingPropertyValueProvider>(StringComparer.OrdinalIgnoreCase);
+            foreach (var valueProvider in valueProviders)
+            {
+                var properyName = valueProvider.Value.GetPropertyName();
```

Comment:
If this is a metadata name, then you don't need to instantiate these unless absolutely needed.